lib/Data/Filesystem.php

```diff
@@ -409,38 +409,47 @@ function ($filename) {
      */
     public function getAllPastes()
     {
-        $pastes  = array();
-        $subdirs = scandir($this->_path);
-        if ($subdirs === false) {
-            dieerr('Unable to list directory ' . $this->_path);
-        }
-        $subdirs = preg_grep('/^[^.].$/', $subdirs);
-
-        foreach ($subdirs as $subdir) {
-            $subpath = $this->_path . DIRECTORY_SEPARATOR . $subdir;
-
-            $subsubdirs = scandir($subpath);
-            if ($subsubdirs === false) {
-                dieerr('Unable to list directory ' . $subpath);
-            }
-            $subsubdirs = preg_grep('/^[^.].$/', $subsubdirs);
-            foreach ($subsubdirs as $subsubdir) {
-                $subsubpath = $subpath . DIRECTORY_SEPARATOR . $subsubdir;
+        $pastes     = array();
+        $firstLevel = array_filter(
+            scandir($this->_path),
+            'PrivateBin\Data\Filesystem::_isFirstLevelDir'
+        );
+        if (count($firstLevel) > 0) {
```

Comment:
code style: here begins a huge block of nested and nested and nested code, so it's very hard for me to follow here. (see also: clean code)

I'd suggest to use early return here: https://dev.to/i3uckwheat/early-return-pattern-558i

Reply:
Also, one may generally extract some functions below into some private functions. E.g. the part in the foreach loop.

Or maybe rewrite this to be recursive (the task of doing that looks like being suitable for that).

Reply:
Also, may I ask the heretical question of what was the aim of this refactoring: just code cleanup (for new PHP versions?) or also performance improvements?

In case of the last one: Did it actually work, do we have benchmarks?
And even if it was not the aim, I hope that such refactoring does at least not _decrease_ the performance. Can we assure that?
(Returning all pastes seems to be a performance relevant step and may be a critical bottleneck, that's just why I am asking…)

Reply:
Purely wanted to match the style of both functions. We can't do an early return, these actually need to loop through all 3 levels of directories to discover all or up to $batchsize * 10 paste IDs. (Aside: that is one area where a RDBMS clearly beats such a naive files on disk storage)

I do like the idea of a recursive function - if we give it a counter argument that it can decrement on each pass, we could use it for both cases. And in any case we should de-dup that logic. I'll work on that, thanks.

Reply:
So thanks for poking me on this. Managed to be a great use case for a glob-based iterator. Instead of us having to filter out things that may not be pastes we now only need to ignore the discussion directories. Much less code!

Some explanation: The glob may return the old paste files without .php suffix, the current ones with suffix and the directory with .discussion suffix. We only need to ignore the latter ones, all others are valid paste files and the ones without suffix will get upgraded by the exists() call. I forgot about that logic, but FilesystemTest::testOldFilesGetConverted() started failing when my pattern did yield only items with the .php suffix.